### PR TITLE
GSoC'20 Menu Implementation

### PR DESCRIPTION
**Pull request by singhalsara8**

### CODE REVIEW

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockComponent.java [hunk 1]
```diff
@@ -641,6 +641,10 @@ public boolean isForm() {
     return false;
   }
 
+  public boolean isMenu() {
+    return false;
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockComponent.java:645]: I recon the changes in this file were from the already implemented solution, right?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockForm.java [hunk 2]
```diff
@@ -88,6 +90,37 @@ public String getTitle() {
...
+        if (child instanceof MockMenu) {
+          menu = (MockMenu) child;
+          break;
+        }
+      }
+      if (menu == null) {
+        menu = new MockMenu(editor);
+        addComponent(menu);
+      }
+      menuButton.addClickHandler(new ClickHandler() {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockForm.java:107]: Interesting, I somehow missed this, but there seems to be something that toggles the actual menu (so making the problem for component overlay solved, potentially?); Have you looked into this, or found something with regards to this?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  private void setTextProperty(String text) {
+    itemWidget.setText(text);
+  }
+
+  /*
+   * Sets the item's Icon property to a new value.
+   */
+  private void setIconProperty(String text) {
+    // Implement this to reflect icon change in designer
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java:72]: This seems like one unimplemented feature; Ideally we should implement it to complete the Menu Options implementation.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  // PropertyChangeListener implementation
+  @Override
+  public void onPropertyChange(String propertyName, String newValue) {
+    super.onPropertyChange(propertyName, newValue);
+
+    // Apply changed properties to the mock component
+    if (propertyName.equals(PROPERTY_NAME_TEXT)) {
+      setTextProperty(newValue);
+      refreshForm();
+    } else if (propertyName.equals(PROPERTY_NAME_ICON)) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java:108]: Do icons work in Android? Have you tried it out?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/palette/SimpleComponentDescriptor.java [hunk 5]
```diff
@@ -10,46 +10,7 @@
...
-import com.google.appinventor.client.editor.simple.components.MockScrollVerticalArrangement;
-import com.google.appinventor.client.editor.simple.components.MockSlider;
-import com.google.appinventor.client.editor.simple.components.MockSpinner;
-import com.google.appinventor.client.editor.simple.components.MockTableArrangement;
-import com.google.appinventor.client.editor.simple.components.MockTextBox;
-import com.google.appinventor.client.editor.simple.components.MockTimePicker;
-import com.google.appinventor.client.editor.simple.components.MockVerticalArrangement;
-import com.google.appinventor.client.editor.simple.components.MockVideoPlayer;
-import com.google.appinventor.client.editor.simple.components.MockWebViewer;
+import com.google.appinventor.client.editor.simple.components.*;
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/palette/SimpleComponentDescriptor.java:13]: Make sure to revert to individual imports; IntelliJ is quite rough because it automatically coerces to `*`, but it's good practice to avoid this. You can generally set an option in the IDE to prevent this (for IntelliJ, set wildcard limit to 999)

File: appinventor/components/src/com/google/appinventor/components/runtime/ListView.java [hunk 6]
```diff
@@ -162,6 +162,10 @@ public View getView() {
     return listViewLayout;
   }
 
+  public android.widget.ListView getListView() {
+    return view;
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ListView.java:167]: Make sure to leave a comment as a javadoc on why the extra view method was required

File: appinventor/components/src/com/google/appinventor/components/runtime/PhoneStatus.java [hunk 7]
```diff
@@ -352,20 +352,6 @@ public static boolean getUseWebRTC() {
     return useWebRTC;
   }
 
-  /**
-   * Static function called from ReplForm when settings menu item is chosen.
-   * Triggers the "OnSettings" event iff there is a PhoneStatus component (which
-   * there will be in the Companion App where this is used).
-   */
-  static void doSettings() {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PhoneStatus.java:360]: Did the implementation that you pulled in remove this?

File: appinventor/components/src/com/google/appinventor/components/runtime/ReplForm.java [hunk 8]
```diff
@@ -200,27 +200,12 @@ public void run() {
   // information and providing the "Settings" menu option.
 
   @Override
-  public boolean onCreateOptionsMenu(Menu menu) {
-    // This procedure is called only once.  To change the items dynamically
-    // we would use onPrepareOptionsMenu.
-    super.onCreateOptionsMenu(menu); // sets up the exit and about buttons
-    addSettingsButton(menu);         // Now add our button!
+  public boolean onPrepareOptionsMenu(Menu menu) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ReplForm.java:203]: Were the changes in this file (`ReplForm.java`) done by the implementation that you pulled in?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        rootPanel.setHeight("100%");
+        menuWidget = new AbsolutePanel();
+        menuWidget.setStylePrimaryName("ode-SimpleMockContainer");
+        menuWidget.add(rootPanel);
+
+        initComponent(menuWidget);
+    }
+
+    @Override
+    protected boolean acceptableSource(DragSource source) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java:36]: Are context menu items linked to the context menu in the hierarchy, much like the menu & menu items?

File: appinventor/components/src/com/google/appinventor/components/runtime/OnPrepareOptionsMenuListener.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package com.google.appinventor.components.runtime;
+
+public interface OnPrepareOptionsMenuListener {
+  void onPrepareOptionsMenu(android.view.Menu menu);
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/OnPrepareOptionsMenuListener.java:8]: Despite this being code written before GSoC, we should still document it in the end with Javadoc to explain the interface, for completeness.

File: appinventor/components/src/com/google/appinventor/components/runtime/MenuItem.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+@DesignerComponent(version = YaVersion.MENUITEM_COMPONENT_VERSION,
+    description = "A Menu Item can only be placed inside Menu components. " +
+    "It displays a piece of text specified by the <code>Text</code> property if " +
+    "shown in options menu, or an icon specified by the <code>Icon</code> property " +
+    "if shown on action bar. Additional properties include visibility and enabled/" +
+    "disabled, all of which can be set in the Designer or Blocks Editor. " +
+    "Click event is triggered on user selection.",
+    category = ComponentCategory.USERINTERFACE)
+@SimpleObject
+public final class MenuItem implements Component {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/MenuItem.java:32]: `MenuItem` looks good to me, and more or less complete. The only things to still consider would be:
* Testing `MenuItem`; Test various cases that induce issues with `null`, and any other potential edge cases that could come up. Given that there's no formal tests written for the component, there might be some underlying issues that were simply overlooked.
* Extra styling options; We could consider those, and look into possibilities on what we can do to get around the limitations of `MenuItem`. Styling in the same fashion as the `Label` would be ideal

File: appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import android.view.View;
+
+/**
+ * Listener for distributing the Activity onCreateContextMenu() method to interested
+ * components.
+ *
+ * @author singhalsara48@gmail.com (Sara Singhal)
+ */
+public interface OnCreateContextMenuListener {
+    public void onCreateContextMenu(android.view.ContextMenu menu, View view, android.view.ContextMenu.ContextMenuInfo menuInfo);
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java]: Make sure to add Javadoc to the method as well, especially to explain each parameter in more detail.

File: appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package com.google.appinventor.components.runtime;
+
+import android.view.View;
+
+/**
+ * Listener for distributing the Activity onCreateContextMenu() method to interested
+ * components.
+ *
+ * @author singhalsara48@gmail.com (Sara Singhal)
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java]: I think from what the looks of it the definition of this interface is unneeded. This seems to be an exact duplicate of the already existing interface: https://developer.android.com/reference/android/view/View.OnCreateContextMenuListener

If there's a concrete reason to have a duplicate interface, make sure to document it. Otherwise you can use `View.OnCreateContextMenuListener` directly.

File: appinventor/components/src/com/google/appinventor/components/runtime/Menu.java [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+  /**
+   * Event to handle when the app user selects an item from the options menu.
+   *
+   * @param itemIndex The index of menu item that is selected (hidden items still take up indices).
+   * @param item The menu item component that is selected.
+   */
+  @SimpleEvent(description = "Event raised when user selects an item from the options menu.")
+  public void ItemSelected(int itemIndex, MenuItem item) {
+    EventDispatcher.dispatchEvent(this, "ItemSelected", itemIndex, item);
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/Menu.java:198]: For consistency, `Context Menu` should probably follow the same convention of including `index` and `item`, if not already.

File: appinventor/components/src/com/google/appinventor/components/runtime/Menu.java [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@DesignerComponent(version = YaVersion.MENU_COMPONENT_VERSION,
+    category = ComponentCategory.LAYOUT,
+    description = "Component for options menu (one per Screen) used to hold MenuItems. "
+        + "Menu is located on the action bar and is not accessible in classic theme.",
+    showOnPalette = false)
+@SimpleObject
+public class Menu implements Component, ComponentContainer, OnCreateOptionsMenuListener,
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/Menu.java:31]: `Menu` seems to look more or less complete. I think the things left to look through would be as follows:

* One design decision that I'd propose is to make `About` and `Stop` be `MenuItem` components that get added to the `Menu` by default. I think this would be a more flexible solution in terms of the menu, and the Menu would be more WYSIWYG in general. @ewpatton perhaps you'd have more insight on this design choice?

* Make sure to test many different possibilities of usage of the menu. Since there are no formal unit or integration tests written, it should be tested thoroughly to gain reliability in the code written. All form cases (onCreate, onPrepare, onOptionsItemSelected, onClear) should generally be tested.

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java [hunk 16]
```diff
@@ -2197,56 +2210,43 @@ private void closeApplication() {
...
+  public Menu getMenu() {
+    return menu;
+  }
+
   @Override
   public boolean onCreateOptionsMenu(Menu menu) {
-    // This procedure is called only once.  To change the items dynamically
-    // we would use onPrepareOptionsMenu.
-    super.onCreateOptionsMenu(menu);
-    // add the menu items
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/Form.java:2205]: These removals were in the previous options menu code, right?

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java [hunk 17]
```diff
@@ -2197,56 +2210,43 @@ private void closeApplication() {
   // Configure the system menu to include items to kill the application and to show "about"
   // information
 
+  private Menu menu = null;
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/Form.java]: This variable should be moved to the top of the class

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java [hunk 18]
```diff
@@ -2284,6 +2285,20 @@ private void showAboutApplicationNotification() {
...
 
+  public void getComponent(View component,ContextMenu menu){
+    componentMap.put(component,menu);
+  }
+
+  @Override
+  public void onCreateContextMenu(android.view.ContextMenu menu, View v, android.view.ContextMenu.ContextMenuInfo menuInfo) {
+    super.onCreateContextMenu(menu, v, menuInfo);
+    for (OnCreateContextMenuListener listener: onCreateContextMenuListeners) {
+      if(listener.equals(componentMap.get(v))) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/Form.java:2298]: After looking at this for a while, I think now it makes sense why the map is used. However, I would strongly suggest documenting on why the `componentMap` is needed and is used the way it is.

File: appinventor/components/src/com/google/appinventor/components/runtime/ContextMenuItem.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+import com.google.appinventor.components.annotations.*;
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.common.YaVersion;
+
+@DesignerComponent(version = YaVersion.CONTEXT_MENUITEM_COMPONENT_VERSION,
+        description = "A Context Menu Item can only be placed inside Context Menu components. " ,
+        category = ComponentCategory.USERINTERFACE)
+@SimpleObject
+public class ContextMenuItem implements Component {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ContextMenuItem.java:21]: A lot of this seems to be strongly re-used from `MenuItem`. I would suggest refactoring it to make use of `MenuItem` functionality in the future to adhere to the DRY principle. Probably one `MenuItem` component would suffice, which would have some `if` logic to add to either a `ContextMenu` or a `Menu`. For the `ContextMenu` and `Menu` components, then you could have some common interface or abstract class (maybe even with a generic parameter, if needed) to support the functionality.

File: appinventor/components/src/com/google/appinventor/components/runtime/ContextMenu.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * Container for ContextMenuItems
+ * @author singhalsara48@gmail.com (Sara Singhal)
+ *
+ */
+@DesignerComponent(version = YaVersion.CONTEXT_MENU_COMPONENT_VERSION,
+        category = ComponentCategory.LAYOUT,
+        description = "Component for context menu to hold ContextMenuItems.")
+@SimpleObject
+public class ContextMenu implements Component, ComponentContainer, OnCreateContextMenuListener {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ContextMenu.java]: Looks good so far! See my comment in `ContextMenuItem` for code reusability; Generally we can apply some generalization here to reduce code duplication

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java (hunk 4 above)
Comment (singhalsara8) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockMenuItem.java:108]: Ya I tried the icon property. When the menuitem is set to be displayed on ActionBar, then icon appears else for overflow menu we can have only text.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/palette/SimpleComponentDescriptor.java (hunk 5 above)
Comment (singhalsara8) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/palette/SimpleComponentDescriptor.java:13]: Ya sure. Thanks! I have made changes to the settings.

File: appinventor/components/src/com/google/appinventor/components/runtime/ListView.java (hunk 6 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/ListView.java:167]: Okay. I will do that.

File: appinventor/components/src/com/google/appinventor/components/runtime/PhoneStatus.java (hunk 7 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/PhoneStatus.java:360]: Yes

File: appinventor/components/src/com/google/appinventor/components/runtime/ReplForm.java (hunk 8 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/ReplForm.java:203]: Yes. I didn't do this.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java (hunk 9 above)
Comment (singhalsara8) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java:36]: Yes. ContextMenu container only accepts ContextMenuItems and no other component can be placed inside it.

File: appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java (hunk 13 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/OnCreateContextMenuListener.java]: No there is no specific reason for having a duplicate interface. I will switch to the already existing one. I just skipped it.

File: appinventor/components/src/com/google/appinventor/components/runtime/Menu.java (hunk 14 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/Menu.java:198]: Okay! I will implement this Event the same way.

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java (hunk 16 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/Form.java:2205]: Ya!

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java (hunk 18 above)
Comment (singhalsara8) [reply appinventor/components/src/com/google/appinventor/components/runtime/Form.java:2298]: Okay, I will document it.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.appinventor.client.widgets.dnd.DragSource;
+import com.google.appinventor.components.common.ComponentConstants;
+import com.google.gwt.user.client.ui.AbsolutePanel;
+
+/**
+ * Mock Popup Menu component.
+ *
+ * @author singhalsara48@gmail.com (Sara Singhal)
+ */
+public class MockPopupMenu extends MockContainer {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java:14]: In my opinion, we should separate the popup menu to another branch. It will make reviewing the menu & context menu alone easier, and also any changes that we will do to context menu after deciding on additional design decisions or technical limitations might need to be propagated to popup menu; So in that sense it would be best to keep them a bit separate.

What do you think?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
+package com.google.appinventor.client.editor.simple.components;
+
+import com.google.appinventor.client.editor.simple.SimpleEditor;
+import com.google.appinventor.client.editor.simple.palette.SimplePaletteItem;
+import com.google.appinventor.client.widgets.dnd.DragSource;
+import com.google.appinventor.components.common.ComponentConstants;
+import com.google.gwt.user.client.ui.AbsolutePanel;
+
+/**
+ * Mock Popup Menu component.
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java:10]: Since the popup menu seems similar to other menus in terms of functionality, do you think it would make sense to generalise them a bit more? For now we could hold off of doing this, but in that sense I would recommend to first finish the context menu before working on the **details** of the popup menu; While experimenting around is great, overengineering the popup menu has a bit of a risk of needing to rewrite a lot of the functionality due to the changes done in context menu once it's finished :)

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+     * Create a new Popup Menu component.
+     */
+    public PopupMenu(ComponentContainer container) {
+        this.container = container;
+        items = new ArrayList<>();
+    }
+
+    public void showPopup(View view) {
+        android.widget.PopupMenu popupMenu = null;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:44]: Make sure to comment these cases. (the if statement)

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+    /**
+     * Create a new Popup Menu component.
+     */
+    public PopupMenu(ComponentContainer container) {
+        this.container = container;
+        items = new ArrayList<>();
+    }
+
+    public void showPopup(View view) {
+        android.widget.PopupMenu popupMenu = null;
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:43]: You can put `popupMenu` declaration inside the if statement I believe, since you're never using it outside

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * Container for MenuItems
+ * @author singhalsara48@gmail.com (Sara Singhal)
+ *
+ */
+@DesignerComponent(version = YaVersion.POPUP_MENU_COMPONENT_VERSION,
+        category = ComponentCategory.LAYOUT,
+        description = "Component for Popup menu to hold MenuItems.")
+@SimpleObject
+public class PopupMenu implements Component, ComponentContainer {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:27]: It looks good, but similarly to the ContextMenu, it seems to have quite a bit of similar functionality as the other menu components; As such I'd recommend to keep the popup menu as is for now, and work on it after you finish Context Menu (unless popup menu is meant to be **completely** different from the context menu); So generally, we should avoid adding extra features to popup menu for now, until we somewhat arrive to a more or less finalised design & implementation of the context menu;

Not to say that we shouldn't work on the popup menu, but be careful of adding more complex functionality that is likely to change in the future if the menu/context menu will change :)

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+        } else if (source instanceof SimplePaletteItem) {
+            component = (MockComponent) source.getDragWidget();
+        }
+        return component instanceof MockContextMenuItem;
+    }
+
+    @Override
+    protected void onSelectedChange(boolean selected) {
+        if (selected) {
+            setVisible(true);
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java]: You can simplify the if statement to one line: `setVisible(selected)`

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+            component = (MockComponent) source;
+        } else if (source instanceof SimplePaletteItem) {
+            component = (MockComponent) source.getDragWidget();
+        }
+        return component instanceof MockContextMenuItem;
+    }
+
+    @Override
+    protected void onSelectedChange(boolean selected) {
+        if (selected) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenu.java]: Make sure to leave a comment on why this is needed

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenuItem.java [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+        if (propertyName.equals(PROPERTY_NAME_TEXT)) {
+            setTextProperty(newValue);
+            refreshForm();
+        }
+    }
+
+    @Override
+    protected void onSelectedChange(boolean selected) {
+        if (selected) {
+            getContainer().setVisible(true);
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContextMenuItem.java:79]: Ditto comment on why this is required

File: appinventor/appengine/war/static/css/Ya.css [hunk 29]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (elatoskinas) [inline appinventor/appengine/war/static/css/Ya.css]: Remove newline at start of file;

File: appinventor/components/build.xml [hunk 30]
```diff
@@ -152,6 +152,7 @@
     <copy toFile="${public.deps.dir}/commons-pool.jar" file="${lib.dir}/commons-pool/commons-pool2-2.0.jar" />
     <copy toFile="${public.deps.dir}/guava-14.0.1.jar" file="${lib.dir}/guava/guava-14.0.1.jar" />
     <copy toFile="${public.deps.dir}/core.jar" file="${lib.dir}/QRGenerator/core.jar" />
+
```
Comment (elatoskinas) [inline appinventor/components/build.xml]: I guess we can remove this change

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockHVLayoutBase.java [hunk 31]
```diff
@@ -395,13 +399,29 @@ private void layoutVertical(LayoutInfo containerLayoutInfo) {
...
 
     for (MockComponent child : containerLayoutInfo.visibleChildren) {
-      dividerLocations[index] = topY;
-      topY += COMPONENT_SPACING;
-
       LayoutInfo childLayoutInfo = containerLayoutInfo.layoutInfoMap.get(child);
       int childWidthWithBorder = childLayoutInfo.width + BORDER_SIZE;
       int childHeightWithBorder = childLayoutInfo.height + BORDER_SIZE;
 
+      if (child.isMenu()) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockHVLayoutBase.java:406]: IMO structuring the 3 subsequent branches as:

`if`, `else if` and `else` is a bit more readable

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContainer.java [hunk 32]
```diff
@@ -171,7 +171,13 @@ private void addComponent(MockComponent component, int beforeIndex) {
       // NOTE: The order of widgets in the root panel does not necessarily
       //       match the order of their associated children of this container
       rootPanel.add(component);
-      refreshForm();
+      refreshForm(true);
+      if (component instanceof MockContextMenu) {
+        component.setVisible(true);
+      }
+      if (component instanceof  MockContextMenuItem) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContainer.java:178]: Ok for now, but if you have the time maybe you could look in if it's possible to re-implement this so that we don't affect `MockContainer`? It would help for separation of concerns, as people modifying `MockContextMenu` in the future might be confused that some of the functionality is actually controlled in `MockContainer` as well

File: appinventor/components/src/com/google/appinventor/components/runtime/AndroidViewComponent.java [hunk 33]
```diff
@@ -75,6 +75,18 @@ public void Visible(boolean visibility) {
...
 
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_COMPONENT + ":com.google.appinventor.components.runtime.ContextMenu")
+  @SimpleProperty
+  public void ContextMenuSelector(ContextMenu menu) {
+    container.$form().getComponent(this.getView(),menu);
+    if(this instanceof ListView) {
+      container.$form().registerForContextMenu(((ListView) this).getListView());
+    }
+    else {
+      container.$form().registerForContextMenu(this.getView());
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/AndroidViewComponent.java:86]: I'd recommend simplifying:
* Define variable `view` for component/view; Set it to `this.getView()` or `(...).getListView()`
* Call `container.$form().registerForContextMenu(view)`

It will be easier to extend if more conditions are added :)

File: appinventor/components/src/com/google/appinventor/components/runtime/AndroidViewComponent.java [hunk 34]
```diff
@@ -75,6 +75,18 @@ public void Visible(boolean visibility) {
     getView().setVisibility(visibility ? View.VISIBLE : View.GONE);
   }
 
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_COMPONENT + ":com.google.appinventor.components.runtime.ContextMenu")
+  @SimpleProperty
+  public void ContextMenuSelector(ContextMenu menu) {
+    container.$form().getComponent(this.getView(),menu);
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/AndroidViewComponent.java:81]: Is this line relevant? Seems like it does not do anything

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 35]
```diff
@@ -0,0 +1,119 @@
...
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.common.YaVersion;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Container for MenuItems
+ * @author singhalsara48@gmail.com (Sara Singhal)
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:20]: We should write a more extensive description, especially how is this different from ContextMenu.

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 36]
```diff
@@ -0,0 +1,119 @@
...
+    private List<ContextMenuItem> items;
+
+    /**
+     * Create a new Popup Menu component.
+     */
+    public PopupMenu(ComponentContainer container) {
+        this.container = container;
+        items = new ArrayList<>();
+    }
+
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:41]: Make sure to add JavaDoc for all the non-override methods below

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 37]
```diff
@@ -0,0 +1,119 @@
...
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR)
+    public AndroidViewComponent ComponentSelector() {
+        return component;
+    }
+
+    @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_COMPONENT )
+    @SimpleProperty
+    public void ComponentSelector(AndroidViewComponent component) {
+        this.component=component;
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:65]: Why is popup menu only allowed on a single component? Shouldn't we allow multiple components to use it?

File: appinventor/components/src/com/google/appinventor/components/runtime/ContextMenu.java [hunk 38]
```diff
@@ -0,0 +1,111 @@
...
+import com.google.appinventor.components.annotations.SimpleProperty;
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.common.YaVersion;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Container for ContextMenuItems
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ContextMenu.java:22]: We should have the Javadoc be more descriptive, especially describing why ContextMenu is different from PopupMenu and the Menu.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java (hunk 21 above)
Comment (elatoskinas) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockPopupMenu.java:14]: By the way, how does the popup menu look like in terms of positioning?

File: appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java [hunk 39]
```diff
@@ -0,0 +1,119 @@
...
+import com.google.appinventor.components.common.YaVersion;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Container for MenuItems
+ * @author singhalsara48@gmail.com (Sara Singhal)
+ *
+ */
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/PopupMenu.java:22]: Could you attach a pic of the differences between ContextMenu vs PopupMenu in Android, and in the mock?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockForm.java (hunk 2 above)
Comment (elatoskinas) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockForm.java:107]: This is resolved I believe, feel free to resolve the comment

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockComponent.java (hunk 1 above)
Comment (singhalsara8) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockComponent.java:645]: Yes
